_posts/contests/2021-02-05-Performance-analysis-and-tuning-contest-4.md

```diff
@@ -312,6 +312,237 @@ One contestant noticed that it is convenient to keep the arrays `delta_x` and `d
+
+                    /* Right point */
+                    z1 = *(magptr - ncols);
+                    z2 = *(magptr - ncols - 1);
+
+                    mag2 = (z2 - z1)*xperp + (m00 - z1)*yperp;
+                }
+            }
+        } 
+
+        /* Now determine if the current point is a maximum point */
+
+        if ((mag1 > 0.0) || (mag2 > 0.0))
+        {
+            *resultptr = (unsigned char) NOEDGE;
+        }
+        else
+        {    
+            if (mag2 == 0.0)
+                *resultptr = (unsigned char) NOEDGE;
+            else
+                *resultptr = (unsigned char) POSSIBLE_EDGE;
+        }
+    }
+  }
+```
+
+The above code is quite verbose, but can be basically divided into three parts. The first part loads the common values `m0`, `gx`, `gy`, `xperp` and `yperp`. The second part on lines TODO performs the calculation dependiong on the values of `gx` and `gy`. The third part writes into the resulting array depending on the calculated values `mag1` and `mag2`.
+
+This nested loop suffers from a large amount of branch mispredictions, which can be confirmed with `perf`. Any attemp in reducing the number of branches will result in better performance. Branches in this case prevent auto-vectorizatiotion, since compilers do not know how to deal with them.
+
+One of the way to reduce number of branches in case of nested branches is to replace nested branches with conditionally joined expression. E.g:
+
+```cpp
+if (gx > 0) {
+  if (gy > 0) {
+    if (gx >= 0) {
+      ...
+    }
+  }
+}
+```
+
+Can be replaced with:
+
+```cpp
+if (gx > 0 & gy > 0 & gx >= 0) {
+  ...
+}
+```
+
+The above three branches are not equivalent. The first example, condition `gy > 0` will be evaluated only if `gx > 0`. In the optimized version, all three conditions are evaluated before the branch is evaluated as a whole. When condition evaluation is cheap (which is normally the case with simple artihmetic conditions), the second solution will be faster. Please note that we are joining the condition with arithmetic `&`, not `&&`, since only when they are joined with `&` will they be evaluated all together.
+
+The above approach will decrease the number of branches, but not eliminate them. We can go completely branches by replacing them with arithmetics. Consider this part of the code:
+
+```cpp
+#define NOEDGE 255
+#define POSSIBLE_EDGE 128
+
+if ((mag1 > 0.0) || (mag2 > 0.0))
+{
+    *resultptr = (unsigned char) NOEDGE;
+}
+else
+{    
+    if (mag2 == 0.0)
+        *resultptr = (unsigned char) NOEDGE;
+    else
+        *resultptr = (unsigned char) POSSIBLE_EDGE;
+}
+```
+
+We conditionally load to `*resultptr` depending on the set of conditions. We load `NOEDGE` if condition `((mag > 0.0) || (mag2 >= 0.0))`, otherwise we load `POSSIBLE_EDGE`. Or to say it the other way around, we load `POSSIBLE_EDGE` if condition `((mag <= 0.0) & (mag2 < 0.0))` is 1. So we can write:
+
+```cpp
+`*resultptr` = POSSIBLE_EDGE *((mag <= 0.0) & (mag2 < 0.0));
```

Comment:
What about storing NOEDGE? We should be storing it to `*resultptr` if `((mag <= 0.0) & (mag2 < 0.0))` is 0, right?

Reply:
I'm not sure if compiler will generate CMOVs for it, but I believe we should use your version from the last comment, as the one in the summary is not valid.

Reply:
Please review this commit: https://github.com/dendibakh/dendibakh.github.io/pull/2/commits/6c66f519a568f84f737fb5a7f762085b63fa1332